libs/sdk-backend-bear/src/backend/workspace/dashboards/widgetFilters.ts

```diff
@@ -0,0 +1,104 @@
+// (C) 2020 GoodData Corporation
+
+import { dashboardFilterReferenceObjRef, IWidget } from "@gooddata/sdk-backend-spi";
+import {
+    filterObjRef,
+    IAttributeFilter,
+    IDateFilter,
+    IFilter,
+    isAllTimeDateFilter,
+    isAttributeFilter,
+    isDateFilter,
+    ObjRef,
+} from "@gooddata/sdk-model";
+import last from "lodash/last";
+import zip from "lodash/zip";
+
+type ObjRefNormalizer = (refs: ObjRef[]) => Promise<string[]>;
+
+/**
+ * Performs widget filter resolution:
+ * - removes all attribute filters hit by ignoreDashboardFilters
+ * - removes all date filters for date dimension different than dateDataSet
+ * - picks the last date filter for the dateDataSet dimension
+ *   - if it is all time, removes all date filters
+ *   - otherwise returns the last date filter specified
+ *
+ * @param widget - widget to resolve filters for
+ * @param filters - filters to try
+ * @param objRefNormalizer - function providing canonical ref value
+ * @internal
+ */
+export async function resolveWidgetFilters(
+    filters: IFilter[],
+    ignoreDashboardFilters: IWidget["ignoreDashboardFilters"],
+    dateDataSet: IWidget["dateDataSet"],
+    objRefNormalizer: ObjRefNormalizer,
+): Promise<IFilter[]> {
+    const dateFilters = filters.filter(isDateFilter);
+    const attributeFilters = filters.filter(isAttributeFilter);
+
+    const isIgnorableFilter = (obj: unknown): obj is IDateFilter | IAttributeFilter =>
+        isDateFilter(obj) || isAttributeFilter(obj);
+
+    if (!dateFilters.length && !attributeFilters.length) {
+        return filters;
+    }
+
+    const [dateFiltersToKeep, attributeFiltersToKeep] = await Promise.all([
+        getRelevantDateFiltersForWidget(dateFilters, dateDataSet, objRefNormalizer),
+        getRelevantAttributeFiltersForWidget(attributeFilters, ignoreDashboardFilters, objRefNormalizer),
+    ]);
+
+    const filtersToKeep = [...dateFiltersToKeep, ...attributeFiltersToKeep];
+
+    // filter the original filter array to maintain order of the items
+    return filters.filter((filter) => !isIgnorableFilter(filter) || filtersToKeep.includes(filter));
+}
+
+async function getRelevantDateFiltersForWidget(
+    filters: IDateFilter[],
+    dateDataSet: IWidget["dateDataSet"],
+    objRefNormalizer: ObjRefNormalizer,
+): Promise<IDateFilter[]> {
+    if (!dateDataSet || !filters.length || filters.every(isAllTimeDateFilter)) {
+        return [];
+    }
+
+    const [dateDatasetUri, ...filterUris] = await objRefNormalizer([
```

Comment:
Very good point, I will rework this!

Reply:
On second glance, I think leaving URIs in is ok. The function is an internal utility and I can be sure it will be working with URIs. Adding the complexity of general ObjRefs is in my opinion not worth it (it could allow some optimization in a sense "if every ref in question is an idRef, do not normalize", however, we know that the inputs here for the ignoreDashboardFilters will always use URIs on bear, so any optimization like this is not feasible, and converting uriRefs to URIs is "for free").

However, I updated the naming to more precisely reflect what we are working with.